fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OutputPropertyDeriver.java

```diff
@@ -0,0 +1,576 @@
+                new HashDistributionDesc(rightOnPredicateColumns, HashDistributionDesc.SourceType.SHUFFLE_JOIN));
+
+        // TODO(ywb): check columns satisfy here instead of enforce phase because there need to compute equivalence columns，
+        // we will do this later
+        Preconditions.checkState(leftOnPredicateColumns.size() == rightOnPredicateColumns.size());
+        // Hash shuffle columns must keep same
+        List<Integer> requiredColumns = requiredShuffleDesc.get().getColumns();
+        boolean checkLeft = leftOnPredicateColumns.containsAll(requiredColumns) &&
+                leftOnPredicateColumns.size() == requiredColumns.size();
+        boolean checkRight = rightOnPredicateColumns.containsAll(requiredColumns) &&
+                rightOnPredicateColumns.size() == requiredColumns.size();
+
+        // @Todo: Modify PlanFragmentBuilder to support complex query
+        // Different joins maybe different on-clause predicate order, so the order of shuffle key is different,
+        // and unfortunately PlanFragmentBuilder doesn't support adjust the order of join shuffle key,
+        // so we must check the shuffle order strict
+        if (checkLeft || checkRight) {
+            for (int i = 0; i < requiredColumns.size(); i++) {
+                checkLeft &= requiredColumns.get(i).equals(leftOnPredicateColumns.get(i));
+                checkRight &= requiredColumns.get(i).equals(rightOnPredicateColumns.get(i));
+            }
+        }
+
+        if (checkLeft) {
+            return createPropertySetByDistribution(leftShuffleDistribution);
+        } else if (checkRight) {
+            return createPropertySetByDistribution(rightShuffleDistribution);
+        } else {
+            return PhysicalPropertySet.EMPTY;
+        }
+    }
+
+    private Optional<HashDistributionDesc> getRequiredShuffleJoinDesc() {
+        if (!requirements.getDistributionProperty().isShuffle()) {
+            return Optional.empty();
+        }
+
+        HashDistributionDesc requireDistributionDesc =
+                ((HashDistributionSpec) requirements.getDistributionProperty().getSpec()).getHashDistributionDesc();
+        if (!HashDistributionDesc.SourceType.SHUFFLE_JOIN.equals(requireDistributionDesc.getSourceType())) {
+            return Optional.empty();
+        }
+
+        return Optional.of(requireDistributionDesc);
+    }
+
+    @Override
+    public PhysicalPropertySet visitPhysicalProject(PhysicalProjectOperator node, ExpressionContext context) {
+        Preconditions.checkState(this.childrenOutputProperties.size() == 1);
+        return childrenOutputProperties.get(0);
+    }
+
+    @Override
+    public PhysicalPropertySet visitPhysicalHashAggregate(PhysicalHashAggregateOperator node,
+                                                          ExpressionContext context) {
+        Preconditions.checkState(this.childrenOutputProperties.size() == 1);
+        return childrenOutputProperties.get(0);
```

Comment:
why hash aggregate could return the child output properties?

Reply:
Because hashAggregate not not change the distrubution. the distribuion is same with its child(eg. exchange, olap scan node)